pkg/crds/crds.go

```diff
@@ -0,0 +1,160 @@
+		"spec": map[string]interface{}{
+			"cleanupPolicy": map[string]string{
+				"confirmation": "yes-really-destroy-data",
+			},
+		},
+	}
+)
+
+func DeleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, clusterNamespace string) {
+	err := deleteCustomResources(ctx, clientsets, clusterNamespace)
+	if err != nil {
+		logging.Fatal(err)
+	}
+	logging.Info("done")
+}
+
+func deleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, clusterNamespace string) error {
+	for _, resource := range cephResources {
+		logging.Info(fmt.Sprintf("Getting resources kind %s", resource))
+		items, err := clientsets.ListResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace)
+		if err != nil {
+			if strings.Contains(err.Error(), errorResourceNotFound.Error()) {
+				logging.Warning("the server could not find the requested resource: %s", resource)
+				continue
+			}
+			// log instead of return
+			return err
+		}
+
+		if len(items) == 0 {
+			logging.Warning("resource %s was not found on the cluster", resource)
+			continue
+		}
+
+		for _, item := range items {
+			logging.Info(fmt.Sprintf("removing resource %s: %s", resource, item.GetName()))
+			err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+			if err != nil {
+				if err.Error() == "the server could not find the requested resource" {
+					logging.Info(err.Error())
+					continue
+				}
+				// log instead of return
+				return err
+			}
+
+			itemResource, err := clientsets.GetResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, item.GetName(), clusterNamespace)
+			if err != nil {
+				if err.Error() == "the server could not find the requested resource" {
+					logging.Info(err.Error())
+					continue
+				}
+				// log instead of return
+				return err
+			}
+
+			if itemResource != nil {
+				logging.Info("resource still alive, applying patch...")
+				err = updatingFinalizers(ctx, clientsets, itemResource, resource, clusterNamespace)
+				if err != nil {
+					if err.Error() == "the server could not find the requested resource" {
+						logging.Info(err.Error())
+						continue
+					}
+					// log instead of return
+					return err
+				}
+
+				err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+				if err != nil {
+					if err.Error() == "the server could not find the requested resource" {
+						logging.Info(err.Error())
+						continue
+					}
+					// log instead of return
+					return err
+				}
+			}
+			logging.Info(fmt.Sprintf("resource %s was removed", item.GetName()))
+		}
+	}
+	return nil
+}
+
+func updatingFinalizers(ctx context.Context, clientsets k8sutil.ClientsetsInterface, itemResource *unstructured.Unstructured, resource, clusterNamespace string) error {
```

Comment:
Where are the finalizers removed in this method? Seems like this is only about the cleanup policy for the cephcluster.

Reply:
I updated the code now, by default if the resource still alive after remove the custom resource, the function `updatingFinalizers()` will try to remove any finalizer, and only will change the finalizer for the resource of kind cephcluster, but always will try to apply the patch if the resource was not removed successfully